### PR TITLE
Kafka

### PR DESCRIPTION
**Pull request by NigeWarren**
Adds Kafka as an Event Source and Sink

### CODE REVIEW

File: src/example/java/io/mewbase/example/shopping/FridgeExample.java [hunk 1]
```diff
@@ -90,7 +89,9 @@ public static void main(String[] args) throws Exception {
         fridgeStatusBinder.get("f1").thenAccept( statusDocumentConsumer );
 
         // Shut that door
```
Comment (kings13y) [inline src/example/java/io/mewbase/example/shopping/FridgeExample.java]: Really ? anyway... moving on!

File: src/main/java/io/mewbase/eventsource/EventSink.java [hunk 2]
```diff
@@ -37,28 +38,28 @@ static EventSink instance(Config cfg) {
...
      * associated Exception.
      *
-     * For blocking version see @EventSink:publish
+     * For blocking version see @EventSink:publishSync
      *
      * @param channelName String of the channel name.
      * @param event as a BsonObject.
      */
-    CompletableFuture<BsonObject> publishAsync(String channelName, BsonObject event);
+    CompletableFuture<Long> publishAsync(String channelName, BsonObject event);
```
Comment (kings13y) [inline src/main/java/io/mewbase/eventsource/EventSink.java]: Any reason sync vs async return different types ? (primitive vs boxed)

File: src/main/java/io/mewbase/eventsource/impl/EventDispatcher.java [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+    public EventDispatcher(Function<T, Event> evtTransformer, EventHandler evtHandler) {
+        this.evtTransformer = evtTransformer;
+        this.evtHandler = evtHandler;
+
+        dispFut = Executors.newSingleThreadExecutor().submit( () -> {
+            while (!closing ) {
+                try {
+                    evtHandler.onEvent(boundedBuffer.take());
+                } catch (InterruptedException exp) {
+                    closing = true;
```
Comment (kings13y) [inline src/main/java/io/mewbase/eventsource/impl/EventDispatcher.java]: Keep wondering if we should add/include finally/Cleaners to these things ?

Comment (kings13y) [reply src/main/java/io/mewbase/eventsource/impl/EventDispatcher.java]: IIRC Cleaners might be Java 9 only.. (https://docs.oracle.com/javase/9/docs/api/java/lang/ref/Cleaner.Cleanable.html)

File: src/main/java/io/mewbase/eventsource/EventSink.java (hunk 2 above)
Comment (kings13y) [reply src/main/java/io/mewbase/eventsource/EventSink.java]: actually. I see why. Wondering if it's worth making sync return the boxed type too though ?

File: src/main/java/io/mewbase/eventsource/impl/file/FileEventUtils.java [hunk 4]
```diff
@@ -34,16 +33,24 @@ static long eventNumberFromPath(final Path path) {
...
+                            currentlyMostRecent.set(f.toFile().lastModified());
+                            return true;
+                        } else {
+                            return false;
+                        }
+                    } )
                     .mapToLong(f -> Long.parseLong(f.toFile().getName()))
                     .max()
-                    .orElse(0) + 1l;
+                    .orElse(-1) + 1l;   //
```
Comment (kings13y) [inline src/main/java/io/mewbase/eventsource/impl/file/FileEventUtils.java]: can't .orElse(0l); ?

File: src/main/java/io/mewbase/eventsource/impl/kafka/KafkaEventSink.java [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+    @Override
+    public long publishSync(final String channelName, final BsonObject event) {
+
+            CompletableFuture<Long> fut = publishAsync(channelName,event);
+            kafkaProducer.flush();
+            try {
+                return fut.get(5, TimeUnit.SECONDS);
+            } catch(Exception exp ) {
+                logger.error("Synchronous publ;ish failed to write event " + event, exp);
+                return -1;
```
Comment (kings13y) [inline src/main/java/io/mewbase/eventsource/impl/kafka/KafkaEventSink.java]: minor typo in that error log. Also, this returns -1 as a long and not an int ?

File: src/main/java/io/mewbase/eventsource/EventSink.java (hunk 2 above)
Comment (NigeWarren) [reply src/main/java/io/mewbase/eventsource/EventSink.java]: Agree - All should return boxed

File: src/main/java/io/mewbase/eventsource/impl/EventDispatcher.java (hunk 3 above)
Comment (NigeWarren) [reply src/main/java/io/mewbase/eventsource/impl/EventDispatcher.java]: Or we could look at AutoClosable and the try with resources pattern

File: src/main/java/io/mewbase/eventsource/impl/file/FileEventUtils.java (hunk 4 above)
Comment (NigeWarren) [reply src/main/java/io/mewbase/eventsource/impl/file/FileEventUtils.java]: Finally got to 0 based Events for all impls - this is defo derived from legacy thinking - it probably reduces although this one works.

File: src/main/java/io/mewbase/eventsource/impl/kafka/KafkaEventSink.java (hunk 5 above)
Comment (NigeWarren) [reply src/main/java/io/mewbase/eventsource/impl/kafka/KafkaEventSink.java]: :-)

**Comment by NigeWarren** [comment]
@kings13y - I'll carry the comments into the JMS branch - as you have kindly merged these

**Comment by NigeWarren** [comment]
@kings13y - Ill carry the comments into the JMS branch - as you have kindly merged these

File: src/main/java/io/mewbase/eventsource/EventSink.java (hunk 2 above)
Comment (NigeWarren) [reply src/main/java/io/mewbase/eventsource/EventSink.java]: Oh and the comment is out of date too

Comment (NigeWarren) [reply src/main/java/io/mewbase/eventsource/EventSink.java]: Applied to JMS branch

File: src/main/java/io/mewbase/eventsource/impl/file/FileEventUtils.java (hunk 4 above)
Comment (NigeWarren) [reply src/main/java/io/mewbase/eventsource/impl/file/FileEventUtils.java]: @kings13y Yep orElse(0) works its on JMS branch ready to go back in next PR

**Pull request merged**